### DISCUSSION TITLE
Using mj_inverse to Obtain Contact Forces

### DISCUSSION THREAD

**Discussion by gaoning0109** [google-deepmind/mujoco#2241] [Asking for Help]
### Intro

Hi!

I am a student  at HEU, I use MuJoCo for my research oncontact forces.

### My setup

MuJoCo 3.2.0   C

### My question

My goal is to use the mj_inverse function to obtain contact forces given qpos, qvel, and qacc. Of course, qpos, qvel, and qacc are correctly calculated from my robot's motion model. I obtained the corresponding forces from cfrc_ext, and I know that the contact forces in cfrc_ext are derived from mj_contactForce. My model is quite simple: it's just a rigid body with a free_joint in the world. After trying, the forces obtained from cfrc_ext do not behave correctly when applied to my motion model; the model flies away. At the same time, I also tested using the mj_forward function, this time only providing qpos and qvel. In this case, the forces obtained from frc_ext added to my motion model behave according to physical laws, and I can also obtain contact forces. According to my understanding, I should be using mj_inverse rather than mj_forward. Perhaps my understanding is incorrect, and I look forward to your response.

### Confirmations

- [X] I searched the [latest documentation](https://mujoco.readthedocs.io/en/latest/overview.html) thoroughly before posting.
- [X] I searched previous [Issues](https://github.com/google-deepmind/mujoco/issues) and [Discussions](https://github.com/google-deepmind/mujoco/discussions), I am certain this has not been raised before.

**Comment by yuvaltassa** [comment]
mj_inverse writes into qfrc_inverse, cfrc_ext is something else.
